Jv rox 15995 plop service should report endpoint when process indicator is not found

JoukoVirtanen opened the pull request:
## Description

The processes listening on ports service should report listening endpoints even if there is not matching process indicator. It is better to in these cases provide partial information about the process associated with a listening endpoint than to mislead by not reporting the listening endpoint or any of the available process information.

## Checklist
- [ ] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] Ran e2e tests 200 times
- [ ] Performance testing
- [ ] Load testing

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -61,13 +61,20 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
+	}
 
 	normalizedPLOPs, completedInBatch := normalizePLOPs(portProcesses)
+	allNormalizedPLOPs := append(normalizedPLOPs, completedInBatch...)
```
See https://github.com/stackrox/stackrox/pull/5204 to understand why this and the following change was made.

erthalion replied:
#5204 is about fixing a lookup for process indicators with empty query. Since there is no lookup, why is it needed?

JoukoVirtanen replied:
The query returns an error if it is empty. Without this change we get empty queries in the unit tests.

JoukoVirtanen commented on central/processlisteningonport/store/postgres/full_store.go:
```diff
@@ -34,9 +34,9 @@ type fullStoreImpl struct {
+	"LEFT OUTER JOIN process_indicators proc " +
 	"ON plop.processindicatorid = proc.id " +
-	"WHERE proc.deploymentid = $1 AND plop.closed = false"
+	"WHERE plop.deploymentid = $1 AND plop.closed = false"
```
We need to use the deploymentid from plop as proc may be NULL as a result of the left outer join.

JoukoVirtanen replied:
Try to optimize query.

erthalion replied:
Again, first show there are any concerns exist about this query.

dashrews78 replied:
agree with @erthalion the query isn't horribly complicated though it could involve a lot of data.  But unless you are seeing problems I'm not sure it is worth spending much time on.

JoukoVirtanen commented on proto/storage/process_listening_on_port.proto:
```diff
@@ -57,4 +58,6 @@ message ProcessListeningOnPortStorage {
     // able to find references process in the database
     ProcessIndicatorUniqueKey process = 7;
+
+    string deployment_id = 8 [(gogoproto.moretags) = 'search:"Cluster ID,store"'];
```
deployment_id is needed here as we now need the plop record to have the deployment_id.

erthalion replied:
I guess the name should be `Deployment ID`, fortunately there is a such search category.

JoukoVirtanen replied:
Done, but I am not 100% that the change I made is correct.

JoukoVirtanen commented on proto/storage/process_listening_on_port.proto:
```diff
@@ -57,4 +58,8 @@ message ProcessListeningOnPortStorage {
+
+    //  string deployment_id = 8 [(gogoproto.moretags) = 'search:"Deployment ID,store,hidden" policy:",prefer-parent" sql:"index=hash,fk(Deployment:id),no-fk-constraint,type(uuid)"'];
+    //string deployment_id = 8 [(gogoproto.moretags) = 'search:"Deployment ID,store" sql:"index=hash,type(uuid)"'];
+    string deployment_id = 8 [(gogoproto.moretags) = 'search:"Deployment ID,store" sql:"type(uuid)"'];
```
I am not confidant about this.

dashrews78 replied:
hash index might make sense on that column

erthalion replied:
I'm afraid this is not immediately obvious. Hash index maybe good for highly unique values, but here `deployment_id` will most likely be quite repeating. The docs [[1]] are saying:

> As a result of the overflow cases, we can say that hash indexes are most suitable for unique, nearly unique data or data with a low number of rows per hash bucket. One possible way to avoid problems is to exclude highly non-unique values from the index using a partial index condition, but this may not be suitable in many cases.

As a small and dirty experiment I've created two tables with only one uuid column, one table having a hash index, another a btree. Then retail inserting 300k records with only 3 distinct values (so, a lot of repetition) shows this:

```
-- insert into the table with btree index
=# insert into test_index_btree select
   case
     when i % 3 = 0 then '1b7e7c08-4c69-4189-9e7d-7474c872f973'::uuid
     when i % 3 = 1 then '98034214-0ad9-41da-a6f8-fefe37224c7b'::uuid
     when i % 3 = 2 then '2f75ed5a-fdbc-454a-8a0b-d22c441583d0'::uuid
   end
   from generate_series(1, 300000) i;

INSERT 0 300000
Time: 1722.175 ms (00:01.722)

-- insert into the table with hash index, somehow slower
=# insert into test_index_hash select
   case
     when i % 3 = 0 then '1b7e7c08-4c69-4189-9e7d-7474c872f973'::uuid
	 when i % 3 = 1 then '98034214-0ad9-41da-a6f8-fefe37224c7b'::uuid
	 when i % 3 = 2 then '2f75ed5a-fdbc-454a-8a0b-d22c441583d0'::uuid
   end
   from generate_series(1, 300000) i;

INSERT 0 300000
Time: 14444.927 ms (00:14.445)

-- size of the involved relations and indexes,
-- hash is almost the same size as the table,
-- where btree is smaller.

             relation             |    size
----------------------------------+------------
 public.test_index_hash_data_idx  | 14 MB
 public.test_index                | 13 MB
 public.test_index_btree          | 13 MB
 public.test_index_btree_data_idx | 1928 kB

```

I'm not saying it's a definitive proof, but probably makes sense to go with btree first, while doing perf tests to compare with hash on the real data distributions. 

[1]: https://www.postgresql.org/docs/current/hash-intro.html

dashrews78 replied:
sounds reasonable

JoukoVirtanen commented on qa-tests-backend/src/test/groovy/BaseSpecification.groovy:
```diff
@@ -37,7 +37,7 @@ import spock.lang.Retry
 import spock.lang.Specification
 
-@Retry(condition = { Helpers.determineRetry(failure) })
+//@Retry(condition = { Helpers.determineRetry(failure) })
```
Uncomment before merging

JoukoVirtanen commented on sensor/common/networkflow/manager/manager_impl.go:
```diff
@@ -132,8 +133,15 @@ func (i *processListeningIndicator) toProto(ts timestamp.MicroTS) *storage.Proce
+		proto.CloseTimestamp = ts.GogoProtobuf()
+	}
+
+	log.Infof("In toProto %+v", proto)
```
Remove before merging

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -96,8 +97,11 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
 		indicatorID := getIndicatorIDForPlop(val)
+		log.Infof("indicatorID= %s", indicatorID)
 		plopKey := getPlopKeyFromParts(val.GetProtocol(), val.GetPort(), indicatorID)
+		log.Infof("plopKey= %s", plopKey)
```
Remove new logging before merging.

JoukoVirtanen commented on proto/storage/process_listening_on_port.proto:
```diff
@@ -57,4 +58,8 @@ message ProcessListeningOnPortStorage {
     ProcessIndicatorUniqueKey process = 7;
+
+    //  string deployment_id = 8 [(gogoproto.moretags) = 'search:"Deployment ID,store,hidden" policy:",prefer-parent" sql:"index=hash,fk(Deployment:id),no-fk-constraint,type(uuid)"'];
+    //string deployment_id = 8 [(gogoproto.moretags) = 'search:"Deployment ID,store" sql:"index=hash,type(uuid)"'];
```
Ask Mandar about this.

JoukoVirtanen committed:
X-Smart-Squash: Squashed 7 commits:

ec116e1311 Check scale performance with indicator index off
146a1cce12 Add best efforts ProcessIndicator matching
2cb006fb12 Setting process correctly
8e39321b58 Check if indicatorID is empty in addNewPLOP
75865b0636 Addressed PR review comments. Changed the type of the value of indicatorsMap to bool since it is not being used
4e25c7d19b Added back metric for orphaned plop
abaebbdb9a Added tests. Added logging. Updated comments

JoukoVirtanen committed:
X-Smart-Squash: Squashed 18 commits:

cba5366838 Got TestPLOPAddNoIndicator to pass
5a037506b4 Removed some unneeded logging
3a5fcc259a Fixed typos
8677ddf5b8 Should work when process indicator matches
02061c2857 Fixed unit tests
efc84099e6 Fixed typo
c779ccf800 Set deploymentid to uuid
fd444dc488 Added debugging
9afe60e3a9 Added debugging
f7ce501ed3 Updated e2e tests by removing asserts that are not gauranteed to pass
58ac9879da Unit tests are passing
73ff6789c9 Debugging
c41a31cda6 Adding ClosedTimestamp to plop from sensor
8a940e252d Setting CloseTimestamp to current timestamp if listening endpoint is missing from the new endpoints and previous timestamp is nil
7730fb659f Clean up
0e081a8510 Fixed style
02b1f5322c Using a hash index on deployment_id in process_listening_on_port.proto
212f09f111 Using a btree index on deployment_id in process_listening_on_port.proto

JoukoVirtanen committed:
Unit tests pass after rebasing. Updated comments in unit tests

JoukoVirtanen committed:
Removed commented out code

JoukoVirtanen committed:
Removed unused variable in e2e tests. Added comments to protobuf for user facing API

JoukoVirtanen committed:
README.md

JoukoVirtanen committed:
Minor change to comment

JoukoVirtanen committed:
Removed fixes for sending CloseTimestamp to central

JoukoVirtanen committed:
Fixed style error

JoukoVirtanen committed:
X-Smart-Branch-Parent: jv-investigate-getting-processindicatorid-without-join-and-index

JoukoVirtanen committed:
make proto-generated-srcs

JoukoVirtanen committed:
Merge branch 'jv-investigate-getting-processindicatorid-without-join-and-index' into jv-rox-15995-plop-service-should-report-endpoint-when-process-indicator-is-not-found

erthalion reviewed (approved):
Looks fine. Please update the PR description and the final squash commit to include the reasoning behind this change.

JoukoVirtanen committed:
Updates DeploymentId for plop if needed
